Return whether string from `expect()` is matched and clear buffer for each call

davisdude opened the pull request:
# TLDR

This PR proposes two changes:

1. Changes `expect()` to return 1 if the pattern was found
2. Adds a new command, `reset_buffer()`, to clear the internal circular buffer

# Justification

## Change to `expect()`

I found it useful to detect whether or not the supplied string was actually matched or not, allowing for more flexible behavior based on what's been found. I found this useful for a few scenarios:

1. Detect whether I need to run a command or not
    * E.g. if I'm already logged (and the `login: ` prompt doesn't show up), there's no reason to enter the password.
    * In code, this would look like:
       ```ini
       script = if (1 == expect("login: ", 1)) then send("root\n") end
       ```
1. Dynamically changing behavior based on hardware
    * I swap which device is connected to a specific harness often. I could have basically identical profiles for each device, but I personally find it easier to just say `tio <dev>` rather than `tio <dev-serial-number>`. Since the login prompt for each device has a unique string that it prompts for on login, I can detect which is present and enter the right password based on that.
    * I have the following Lua script for that:
      ```lua
      local logins = {
          {
              serialnumber = "foo",
              username = "foouser",
              password = "foopass",
          },
          {
              serialnumber = "bar",
              username = "baruser",
              password = "barpass",
          },
          {
              serialnumber = "baz",
              username = "bazuser",
              password = "bazpass",
          },
      }
      
      reset_buffer()
      for _, login in ipairs(logins) do
          send("\n")
          local found = expect(login.serialnumber .. ".*login:", 10)
          if (1 == found) then
              send(login.username .. "\n")
              expect("Password:")
              send(login.password .. "\n")
              break
          end
      end
      ```
    * This script is obviously not the most efficient, since it has to check for each device one at a time. If there was a way to determine which pattern in a regex was matched, we could concat all the serial numbers into a single pattern, then proceed based on which was matched.
    * Also probably not ideal to send `\n` each time...

So the return codes for `expect` are now:

* -1: Invalid regex
* 0: Timeout
* 1: Match

## `reset_buffer()`

At first, the matching wasn't quite as "smart" as I'd hoped, since the last 2000 characters are buffered. This meant that it was possible to get false positives in some instances, so I also added a `reset_buffer()` command as well to flush the buffer. After flushing the buffer, the behavior matched what I expected.

# Closing

I'm totally open to this functionality not being desired. I just like Lua and wanted to get this behavior through it if I could, and thought others might find it helpful. I'm a fairly new tio user, so please let me know if this should / could be handled in a better way (probably using sockets?). Also, please let me know if there's documentation / unit tests I should update. I see something similar to this is already in the [TODO](https://github.com/tio/tio/blob/1e20948d83a82b31a058557f6158d0dbd1a2e655/TODO#L35C3-L35C60), so maybe this isn't necessary. I just found it useful for my own needs.

davisdude commented:
Converting to a draft while I troubleshoot an issue when `expect`ing something that isn't immediately matched...

lundmar commented:
The first first change looks good to me. I actually intended it to return 1 on match but clearly I forgot to add that.

Regarding the second change I don't think we need to expose reset_buffer(). Instead we should make expect() behave correctly by moving the code from reset_buffer() to the expect() c function so we reset the circular buffer before each expect session.

lundmar commented:
> I'm totally open to this functionality not being desired. I just like Lua and wanted to get this behavior through it if I could, and thought others might find it helpful. I'm a fairly new tio user, so please let me know if this should / could be handled in a better way (probably using sockets?). Also, please let me know if there's documentation / unit tests I should update. I see something similar to this is already in the [TODO](https://github.com/tio/tio/blob/1e20948d83a82b31a058557f6158d0dbd1a2e655/TODO#L35C3-L35C60), so maybe this isn't necessary. I just found it useful for my own needs.

It's all good. I expect some input on the new script feature so we can work out any kinks and add new useful features.

davisdude commented:
> Regarding the second change I don't think we need to expose reset_buffer().

Reverted in a9cc0fb.

> Instead we should make expect() behave correctly by moving the code from reset_buffer() to the expect() c function so we reset the circular buffer before each expect session.

Updated behavior in 10e6384.

lundmar commented:
Thank you for testing and improving the new script feature.

davisdude commented:
Thanks for tio! It does exactly what I want it to and was super easy to set up.

lundmar commented:
@davisdude I hope you don't mind. I've added your automatic login script to the newly created lua examples directory. Feel free to update or add more scripts if you think it may be of use to others :)

davisdude commented:
I don't mind at all! I'm glad it's a helpful example :slightly_smiling_face:

lundmar commented:
Oh, I just realize that your example would not work anymore.

I think we need some sort of read feature so you can read response strings from the serial device so you can evaluate and act accordingly.

davisdude commented:
I tested and it does work for me, though I have to send a line feed before each expect. A read method would certainly make things more elegant. 

E.g. connecting to the third device in the list:

```shell
$ tio dev
[11:48:55.696] tio v2.9
[11:48:55.697] Press ctrl-t q to quit
[11:48:55.697] Connected
[11:48:55.698] Running script /home/user/.config/tio/dev.lua

baz login: 
baz login: 
baz login: bazuser
Password: 
Last login: Wed Apr 10 19:16:55 +0000 2024 on /dev/ttyPS0.
bazuser@baz:~$

# Running script again with CTRL-T r
[11:52:28.375] Running script /home/user/.config/tio/dev.lua

bazuser@baz:~$
bazuser@baz:~$
bazuser@baz:~$
```

lundmar commented:
Oh right, you use the line feed trick to test different logins.

Yeah, there must be a better way to do this. I'll need to study how expect handles this scenario.

lundmar commented:
I've extended expect() to return the matched string as second parameter. This way, with the right regular expression, it should be possible to make the automatic login script more elegant without having to perform repeated logins.

Instead, now you can do something like this:
```
found,s = expect("\\w+ login:", 100)
```
When match: found = 1, s = "baz login:"

From there it should be trivial to match your serialnumbers using string.find() or similar lua mechanism.

What do you think?

davisdude commented:
That sounds great! Thanks a bunch :slightly_smiling_face:

lundmar commented:
If you get it working feel free to update the example login script.
